README.md
DivSugar
========

**DivSugar is a CSS-based 3D graphics library,** which provides a 3D scene graph, an animation system, and geometry classes.
Each node of the scene graph is an extended div element, so it is very easy to integrate 3D animations into existing web pages.

DivSugar depends on CSS 3D transforms and runs on Safari, Chrome, and Firefox for now.

<a href="http://kitao.github.com/divsugar/examples/01.SimpleApplication/" target="_blank">
  <img src="http://kitao.github.com/divsugar/examples/screenshots/01.SimpleApplication.png" width="19.5%">
</a>
<a href="http://kitao.github.com/divsugar/examples/02.TaskAnimation/" target="_blank">
  <img src="http://kitao.github.com/divsugar/examples/screenshots/02.TaskAnimation.png" width="19.5%">
</a>
<a href="http://kitao.github.com/divsugar/examples/03.TweenAnimation/" target="_blank">
  <img src="http://kitao.github.com/divsugar/examples/screenshots/03.TweenAnimation.png" width="19.5%">
</a>
<a href="http://kitao.github.com/divsugar/examples/04.LightAndBillboard/" target="_blank">
  <img src="http://kitao.github.com/divsugar/examples/screenshots/04.LightAndBillboard.png" width="19.5%">
</a>
<a href="http://kitao.github.com/divsugar/examples/05.KeyAndMouseInput/" target="_blank">
  <img src="http://kitao.github.com/divsugar/examples/screenshots/05.KeyAndMouseInput.png" width="19.5%">
</a>

Getting Started
---------------

Include `divsugar.js`, then all of the functions get available via the `DivSugar` global variable.

```html
<script src="divsugar.js"></script>
```

A Scene is the root of a scene graph and specifies the rendering area.
A Node is an element of a scene graph and draws an image.
Both of them contain div elements as their properties.

```javascript
var scn = new DivSugar.Scene()
  .setSize(800, 600)
  .setImage('#000080')
  .appendTo(document.body);

var node = new DivSugar.Node()
  .setSize(300, 300)
  .setPosition(10, 20, 30)
  .setImage('image.jpg')
  .appendTo(scn);
```

DivSugar provides a Task, which calls the `onUpdate` method of the registered Tasks in every frame.
It also calls the `onDestroy` method when the Task is destroyed.

```javascript
var task = new DivSugar.Task().appendTo(DivSugar.rootTask);

task.onUpdate = function() {
  node.rotate(this.deltaTime * 0.1, this.deltaTime * 0.2, this.deltaTime * 0.3);
};

task.onDestroy = function() {
  scn.removeChild(node);
};
```

With the `playAnimation` method of a Node, it is possible to make the Task which plays a sequence of animations automatically.

```javascript
node.playAnimation([
  ['to', { size: [100, 100], opacity: 0 }, 500, DivSugar.Ease.quadInOut],
  ['wait', 500],
  ['to', { size: [200, 200], opacity: 1 }, 1000, DivSugar.Ease.linear],
  ['play', someAnim],
  ['call', someFunc],
  ['repeat', 10]
]);
```

Examples
--------

- 01.SimpleApplication ([demo](http://kitao.github.com/divsugar/examples/01.SimpleApplication/) | [code](https://github.com/kitao/divsugar/blob/master/examples/01.SimpleApplication))
- 02.TaskAnimation ([demo](http://kitao.github.com/divsugar/examples/02.TaskAnimation/) | [code](https://github.com/kitao/divsugar/blob/master/examples/02.TaskAnimation))
- 03.TweenAnimation ([demo](http://kitao.github.com/divsugar/examples/03.TweenAnimation/) | [code](https://github.com/kitao/divsugar/blob/master/examples/03.TweenAnimation))
- 04.LightAndBillboard ([demo](http://kitao.github.com/divsugar/examples/04.LightAndBillboard/) | [code](https://github.com/kitao/divsugar/blob/master/examples/04.LightAndBillboard))
- 05.KeyAndMouseInput ([demo](http://kitao.github.com/divsugar/examples/05.KeyAndMouseInput/) | [code](https://github.com/kitao/divsugar/blob/master/examples/05.KeyAndMouseInput))

Documentaion
------------

- [API Reference](https://github.com/kitao/divsugar/blob/master/docs)
- Tutorial ([English](https://github.com/kitao/divsugar/wiki/DivSugar-Tutorial) | [Japanese](https://github.com/kitao/divsugar/wiki/DivSugar-Tutorial-in-Japanese))

License
-------

- MIT license


CHANGELOG.md
Change Log
==========

DivSugar 1.2.1
--------------
- Replaced the deprecated APIs of qunit
- Updated the URL of qunit
- Fixed a typo of the example #3

DivSugar 1.2.0
--------------
- Renamed the clearAnimation method of the Node class to clearAnimations
- Minified the library
- Modifed the README.md a little
- Added a tutorial to wiki

DivSugar 1.1.5
--------------
- Changed to use an event listener instead of an event handler
- Modified the example #5 for touch events
- Added support for touch events

DivSugar 1.1.0
--------------
- Added an example of key and mouse input
- Added the rotateAround method to the Matrix and Node classes
- Changed to throw exceptions for unknown states and styles
- Added the getMouseState method to the DivSugar class
- Added the getMouseX and getMouseY methods to the DivSugar class
- Added the getKeyState method to the DivSugar class
- Added the getFrameCount method to the DivSugar class

DivSugar 1.0.5
--------------
- Added the screenshots of the examples to the README.md
- Optimized the toCSSTransform2D/3D method of the Matrix class a little

DivSugar 1.0.0
--------------
- Added the description of all of the APIs
- Added the animation commands and properties to the API docs
- Added the type of the return values to the API docs
- Added support for Opera partly (only use 2D Transforms)
- Added the toCSSTransform2D method to the Matrix class
- Renamed the toCSSTransform method of the Matrix class to toCSSTransform3D

DivSugar 0.9.5
--------------
- Fixed the perspective origin error on Firefox
- Changed the way of cross-browser support
- Removed the browserPrefix property of the DivSugar class

DivSugar 0.9.0
--------------
- Added an example of lighting and billboard
- Modified the way of cross-browser CSS handling for Firefox
- Modified the message for the exceptions
- Added the getCSSColor method to the DivSugar class
- Added the browserPrefix property to the DivSugar class

DivSugar 0.8.0
--------------
- Added the change log
- Added the API docs
- Added the version number constant
- Changed to throw exceptions for unknown animation commands/params
- Changed the rootNode property of the Scene class to private
- Added the getParent methods to the Scene and Node class
- Removed the isRootTask property of the rootTask
- Changed the Scene and Node class to contain a div element
- Removed the createScene method and the createNode method of the DivSugar class
- Integrated the rotate methods of the Vector class
- Changed to throw exceptions for invalid argument number
- Added the getLocalPosition method to the Scene class

DivSugar 0.7.0
--------------
- First release


examples/03.TweenAnimation/main.js
window.onload = function() {
  'use strict';

  // create a Scene
  var scn = new DivSugar.Scene().setSize(800, 600).setImage('#000030').appendTo(document.body);

  // maximize the Scene size
  function resize() { scn.adjustLayout(window.innerWidth, window.innerHeight, 'contain'); }
  window.addEventListener('resize', resize, true);
  resize();

  // create the center Node
  var center = new DivSugar.Node().setPosition(400, 300, 0).appendTo(scn);
  center.playAnimation([['to', { rotate: [120, 0, -10] }, 5000, DivSugar.Ease.cubicInOut], ['wait', 2000], ['repeat']]);

  // create the banner Nodes
  var x, y, z, color, waitTime, moveTime, anim, banner;
  var colors = ['#ffff80', '#ff80ff', '#80ffff', '#ff8080', '#80ff80', '#8080ff'];

  for (var i = 0; i < 40; i++) {
    x = Math.random() * 800 - 800;
    y = Math.random() * 1000 - 500;
    z = Math.random() * 1000 - 500;

    color = colors[Math.floor(Math.random() * 6)];

    waitTime = Math.random() * 5000;
    moveTime = Math.random() * 3000 + 3000;

    anim = [
      ['to', { position: [x + 400, y, z], opacity:   0 }],
      ['to', { position: [x + 200, y, z], opacity: 0.6 }, moveTime / 4],
      ['to', { position: [x - 200, y, z], opacity: 0.6 }, moveTime / 2],
      ['to', { position: [x - 400, y, z], opacity:   0 }, moveTime / 4],
      ['repeat']
    ];

    banner = new DivSugar.Node().setSize(1200, 100).setOpacity(0).rotate(0, 0, Math.random() * 180 - 90).appendTo(center);
    banner.div.innerHTML = '<h1 style="color:' + color + '; font-size:50px">DivSugar, a CSS-based 3D graphics library</h1>';
    banner.playAnimation([['wait', waitTime], ['play', anim]]);
  }
};
